pkg/config/config.go

```diff
@@ -358,6 +358,9 @@ type EngineConfig struct {
 	// OCIRuntimes are the set of configured OCI runtimes (default is runc).
 	OCIRuntimes map[string][]string `toml:"runtimes,omitempty"`
 
+	// PlatformToOCIRuntime requests specific OCI runtime for a specified platform of image.
+	PlatformToOCIRuntime map[string]OCIRuntimeList `toml:"platform_to_oci_runtime,omitempty"`
```

Comment:
Why is it not a simple `map[string]string`?

I imagined something like `foo/bar=crun`. We already have a map for the different paths a runtime may live in. This reduces redundancy.

Reply:
also a similar ack was here https://github.com/containers/common/pull/1225#issuecomment-1307052541

Reply:
I am fine with simplifying it to map[string]string.
@giuseppe WDYT?

Reply:
`map[string]string` prohibits from adding more field in future, I think in future we might wanna add more config fields like which `WASM` runtime to use if multiple are present on the host. Same goes for other handlers.

Reply:
Give me an example use case?

Reply:
There are lot of things for WASI capabilities which all handlers are doing by default  for example pre-opening `workdir` dir, initially i thought that these must be part of image annotations and configurable but since standard is going towards platform string so not sure if we should still use annotations . Maybe if we make it configure-able in future I think following struct can include extra configuration field. This was also my concern when feature was implemented https://github.com/containers/crun/pull/985#issuecomment-1211910701

And all capabilities config can be centrally maintained in `container.conf`, but again there is no request from users as of now but I think in future there might be.

```toml
[engine.platform_to_oci_runtime]
"wasi/wasm" = { wasi_capablities = ["preopen_workdir:true", "preopen_sockets:true"],  runtimes = ["/usr/bin/crun-wasm"] }
"wasi/wasm32" = { runtimes = ["/usr/bin/crun-wasm"] }
"wasi/wasm64" = { runtimes = ["/usr/bin/crun-wasm"] }
```

Reply:
one issue we have with having just a path now is that additional information, e.g. we already have `runtime_supports_json`, `runtime_supports_kvm` and `runtime_supports_nocgroups`, while this information could be part of the runtime struct itself

Reply:
Regarding **looking up paths** we have [[engine.runtimes]](https://github.com/containers/common/blob/main/pkg/config/containers.conf#L615) already. So I am very much against duplicating that for `[engine.platform_to_oci_runtime]`.  The mapping IMO should only be `os/arch->name` where `name` can/must further be resolved by consulting the `[engine.runtimes]`.  That will reduce redundancy.

Regarding **extending/adding attributes of the runtimes**: as Giuseppe mentioned, we have extra fields/options already.  It would be nicer to have a more complex and _centralized_ structure for that.  But I don't think that `[engine.platform_to_oci_runtime]` is the right place for that.  As the name indicates, I think it should really only be a translation/mapping from `os/arch -> name`.  Adding more attributes is mixing concerns and can cause issues down the road. 

Let's assume we need an attribute for a non-wasm case: Would we have to add the same redundant entry for all `linux/{amd64,aarch64,arm,ppc,s390,...}`? It doesn't look right.

Reply:
I have a different perspective towards `[engine.runtimes]` I think they should not be mixed with the use-case specific runtimes. 

I think `[engine.runtimes]` is a field for OCI runtimes. While all these modern handlers are not OCI runtimes they are just special modification which allows users to wrap certain workload in OCI images, `runtime-spec` does not talks about these special modifications so mixing these with `[engine.runtimes]` sets a confusing perspective for users.

My idea was to never mix `platform->runtime` with OCI runtime in `[engine.runtimes]` but its just my opinion and I'm open to suggestion.

Reply:
I think individual fields `runtime_supports_kvm` , `runtime_supports_kvm` does not fits well with the use-case mentioned here https://github.com/containers/common/pull/1225#discussion_r1018658462

Reply:
Overall I think it should be okay to drop struct for now as use `map[string]string` right ? if that's the consensus and we can add it later or think of a central config for configuration later on when strong use-case appears.

@giuseppe @vrothberg WDYT ?

Reply:
`[engine.runtimes]` is used to lookup runtimes at different paths, primarily allowing to support multiple distributions which have varying installation paths and conventions.

In it's current state, the PR duplicates that and possibly for each entry.  It should be `1->N` mapping, not a `M->N`.

If we want to make something specific to WASM, I am not against it but `[engine.platform_to_oci_runtime]` very much suggests we're attempting something generic.  "platform_to_wasm_runtime" may be better suited.

In general I try to enforce separation of concerns and look for generalizing use cases.  While it's sometimes better to specialize (instead of generalizing) things, I take separation of concerns as something dogmatic; mostly because mixing concerns tends to lead to one-way-roads.

Maybe we can take a step back and isolate which problems we want to solve.  The PR started with mapping `paltform->runtime` and grew a bit.  What else is on the table?

Reply:
> What else is on the table?

@vrothberg I don't see lot of use-cases in hand but its only speculation for future. Overall on thinking again I like  the apporach you have shared here https://github.com/containers/common/pull/1225#discussion_r1018804461

`platform_to_oci_runtime` is `map[string]string`: and only points to the runtime for a platform
and actual paths is defined in `[engine.runtimes]`

For runtime specific configurations we can define a central config struct later on when strong use-case appears in future.

Reply:
Thanks, @flouthoc!

> For runtime specific configurations we can define a central config struct later on when strong use-case appears in future.

:100: that will allow for consolidating the other fields as well that @giuseppe pointed out.

Reply:
@giuseppe @rhatdan I'll amend this as per suggestions above if there is a consensus on this one.

Reply:
SGTM